Tests/MeshLib/TestAddLayerToMesh.cpp

```diff
@@ -0,0 +1,232 @@
+	// created during the extrusion. Therefore node order tests are switched off for line meshes.
+	void validate(MeshLib::Mesh const& mesh, bool testNodeOrder)
+	{
+		int reduce_tests = (testNodeOrder) ? 0 : 1;
+
+		const std::size_t nErrorFlags (static_cast<std::size_t>(ElementErrorFlag::MaxValue));
+		ElementErrorFlag flags[nErrorFlags] = {ElementErrorFlag::ZeroVolume, 
+		ElementErrorFlag::NonCoplanar, ElementErrorFlag::NonConvex,  ElementErrorFlag::NodeOrder};
+		std::vector<ElementErrorCode> codes (MeshLib::MeshValidation::testElementGeometry(mesh));
+		for (std::size_t i=0; i<codes.size(); ++i)
+			for (std::size_t j=0; j<nErrorFlags-reduce_tests; ++j)
+				ASSERT_EQ(false, codes[i][flags[j]]);
+	}
+
+	void testZCoords2D(MeshLib::Mesh const& input, MeshLib::Mesh const& output, int height)
+	{
+		std::size_t nNodes (input.getNNodes());
+		for (std::size_t i=0; i<nNodes; ++i)
+		{
+			ASSERT_EQ((*input.getNode(i))[2], (*output.getNode(i))[2]);
+			ASSERT_EQ((*input.getNode(i))[2] + height, (*output.getNode(nNodes+i))[2]);
+		}
+	}
+
+	void testZCoords3D(MeshLib::Mesh const& input, MeshLib::Mesh const& output, int height)
+	{
+		std::size_t nNodes (input.getNNodes());
+		for (std::size_t i=0; i<nNodes; ++i)
+			ASSERT_EQ((*input.getNode(i))[2] + height, (*output.getNode(i))[2]);
+	}
+};
+
+TEST(MeshLib, AddLayerToLineMesh)
+{
+	std::unique_ptr<MeshLib::Mesh> mesh (MeshLib::MeshGenerator::generateLineMesh(1.0, 5));
+
+	// add top layer
+	int height (1);
+	std::unique_ptr<MeshLib::Mesh> result (MeshLib::addLayerToMesh(*mesh, height, "mesh", true));
+
+	ASSERT_EQ(2*mesh->getNNodes(), result->getNNodes());
+	ASSERT_EQ(2*mesh->getNElements(), result->getNElements());
+
+	std::array<unsigned, 7> n_elems (MeshLib::MeshInformation::getNumberOfElementTypes(*result));
+	ASSERT_EQ(5, n_elems[0]);
+	ASSERT_EQ(5, n_elems[2]);
+
+	AddLayerValidation::testZCoords2D(*mesh, *result, height);
+	AddLayerValidation::validate(*result, false);
+
+	// add bottom layer
+	result.reset(MeshLib::addLayerToMesh(*mesh, height, "mesh", false));
+
+	ASSERT_EQ(2*mesh->getNNodes(), result->getNNodes());
+	ASSERT_EQ(2*mesh->getNElements(), result->getNElements());
+
+	n_elems = MeshLib::MeshInformation::getNumberOfElementTypes(*result);
+	ASSERT_EQ(5, n_elems[0]);
+	ASSERT_EQ(5, n_elems[2]);
+
+	AddLayerValidation::testZCoords2D(*mesh, *result, -1 * height);
+	AddLayerValidation::validate(*result, false);
+}
+
+TEST(MeshLib, AddLayerToTriMesh)
+{
+	std::unique_ptr<MeshLib::Mesh> mesh (MeshLib::MeshGenerator::generateRegularTriMesh(5, 5));
+
+	// add top layer
+	int height (1);
+	std::unique_ptr<MeshLib::Mesh> result (MeshLib::addLayerToMesh(*mesh, height, "mesh", true));
+
+	ASSERT_EQ(2*mesh->getNNodes(), result->getNNodes());
+	ASSERT_EQ(2*mesh->getNElements(), result->getNElements());
+
+	std::array<unsigned, 7> n_elems (MeshLib::MeshInformation::getNumberOfElementTypes(*result));
+	ASSERT_EQ(50, n_elems[1]);
```

Comment:
Can you add a comment what is tested here, i.e., what is `n_elems[1]` and `n_elems[6]` stands for?